numpy/fft/fftpack.py

```diff
@@ -56,10 +59,13 @@ def _raw_fft(a, n=None, axis=-1, init_function=fftpack.cffti,
                          % n)
 
     try:
-        # Thread-safety note: We rely on list.pop() here to atomically
-        # retrieve-and-remove a wsave from the cache.  This ensures that no
-        # other thread can get the same wsave while we're using it.
-        wsave = fft_cache.setdefault(n, []).pop()
+        # We have to ensure that only a single thread can access a wsave array
+        # at any given time. Thus we remove it from the cache and insert it
+        # again after it has been used. Multiple threads might create multiple
+        # copies of the wsave array. This is intentional and a limitation of
+        # the current C code.
+        with _cache_lock:
```

Comment:
I thought about this but decided against it in the end for a couple of reasons:

(1) It makes the code slightly awkward - the locks are members on the cache objects but are acquired and released outside of it. In that case I'd prefer the previous solution where the cache managed its lock.

(2) Performance wise it has no measurable effect as all the operations in the lock are super quick. Also use cases where both real and complex FFTs are called a lot in a threaded context are probably rare.

But its your call. Let me know.